packages/workbox-runtime-caching/NetworkFirst.mjs

```diff
@@ -0,0 +1,162 @@
+   * requests. Defaults to cache names provided by `workbox-core`.
+   * @param {string} options.plugins Workbox plugins you may want to use in
+   * conjunction with this caching strategy.
+   * @param {number} options.networkTimeoutSeconds If set, any requests that
+   * fail to respond within this timeout will be treated as a failed network
+   * request and the request will fallback to the cache. This option is meant
+   * to combat "[lie-fi](https://developers.google.com/web/fundamentals/performance/poor-connectivity/#lie-fi)"
+   * scenarios.
+   */
+  constructor(options = {}) {
+    this._cacheName =
+      _private.cacheNames.getRuntimeName(options.cacheName);
+
+    if (options.plugins) {
+      let isUsingCacheWillUpdate = false;
+      for (let plugin of options.plugins) {
+        if (plugin.cacheWillUpdate) {
+          isUsingCacheWillUpdate = true;
+          break;
+        }
+      }
+      this._plugins = isUsingCacheWillUpdate ?
+        options.plugins : [defaultPlugin, ...options.plugins];
+    } else {
+      // No plugins passed in so just use default plug.
+      this._plugins = [defaultPlugin];
+    }
+
+    this._networkTimeoutSeconds = options.networkTimeoutSeconds;
+    if (process.env.NODE_ENV !== 'production') {
+      if (this._networkTimeoutSeconds) {
+        core.assert.isType(this._networkTimeoutSeconds, 'number', {
+          moduleName: 'workbox-runtime-caching',
+          className: 'NetworkFirst',
+          funcName: 'constructor',
+          paramName: 'networkTimeoutSeconds',
+        });
+      }
+    }
+  }
+
+  /**
+   * Handle the provided fetch event.
+   *
+   * @param {FetchEvent} event
+   * @return {Promise<Response>}
+   */
+  async handle(event) {
+    if (process.env.NODE_ENV !== 'production') {
+      // TODO: Switch to core.assert
+      // core.assert.isInstance({event}, FetchEvent);
+    }
+
+    const promises = [];
+    let timeoutId;
+
+    if (this._networkTimeoutSeconds) {
+      promises.push(new Promise((resolve) => {
+        const onNetworkTimeout = () => {
+          resolve(this._respondFromCache(event.request));
+        };
+
+        timeoutId = setTimeout(
+          onNetworkTimeout,
+          this._networkTimeoutSeconds * 1000,
+        );
+      }));
+    }
+
+    const networkPromise = _private.fetchWrapper.fetch(
+      event.request,
+      this._plugins
+    )
+    .then((response) => {
+      if (timeoutId) {
+        clearTimeout(timeoutId);
+      }
+
+      if (!response) {
+        return this._respondFromCache(event.request);
+      }
+
+      // Keep the service worker alive while we put the request in the cache
+      const responseClone = response.clone();
+      event.waitUntil(
+        _private.cacheWrapper.put(
+          this._cacheName,
+          event.request,
+          responseClone,
+          this._plugins
+        )
+      );
+
+      return response;
+    }, () => this._respondFromCache(event.request));
```

Comment:
My preference is for explicit `.catch()` rather than the two-argument `.then()`.